### PR TITLE
[CosmosDB]  Managed Cassandra

### PR DESCRIPTION
**Pull request by AverageDesigner**
In this PR, we are providing support through az cli for the new service ManagedCassandra.

Managed Cassandra is a new service that is positioned as a separate product but operated as an extension of CosmosDB.

In Managed Cassandra the top-level resource is a cluster. each cluster can have multiple data centers.

We are using cassandra-mi as the az command for managed cassandra.

### CODE REVIEW

**Review by kavskalyan — CHANGES_REQUESTED** [review]

**Comment by yonzhan** [comment]
CosmosDB

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 1]
```diff
@@ -139,6 +139,55 @@ def load_arguments(self, _):
         c.argument('account_name', account_name_type, id_part=None)
         c.argument('role_assignment_id', options_list=['--role-assignment-id', '-i'], validator=validate_role_assignment_id, help="Optional for Create. Unique ID for the Role Assignment. If not provided, a new GUID will be used.")
         c.argument('role_definition_id', options_list=['--role-definition-id', '-d'], validator=validate_fully_qualified_role_definition_id, help="Unique ID of the Role Definition that this Role Assignment refers to.")
-        c.argument('role_definition_name', options_list=['--role-definition-name', '-n'], help="Unique Name of the Role Definition that this Role Assignment refers to. Eg. 'Contoso Reader Role'.")
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:142]: Why is this removed ? It might be unintentional because your branch might not be synced with master.

File: src/cosmosdb-preview/azext_cosmosdb_preview/commands.py [hunk 2]
```diff
@@ -15,10 +15,13 @@
...
     cf_restorable_mongodb_collections,
     cf_restorable_mongodb_resources,
-    cf_sql_resources
+    cf_sql_resources,
+    cf_cassandra_cluster,
+    cf_cassandra_data_center
 )
 
 
+
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/commands.py:24]: nit: remove extra line

File: src/cosmosdb-preview/azext_cosmosdb_preview/custom.py [hunk 3]
```diff
@@ -451,7 +454,7 @@ def cli_cosmosdb_sql_role_definition_exists(client,
     """Checks if an Azure Cosmos DB Sql Role Definition exists"""
     try:
         client.get_sql_role_definition(role_definition_id, resource_group_name, account_name)
-    except CloudError as ex:
+    except C as ex:
```
[the remark this answers is not in the document]
Comment (MehaKaushik) [reply src/cosmosdb-preview/azext_cosmosdb_preview/custom.py:457]: following this conversation

File: src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/_cosmos_db_management_client.py [hunk 4]
```diff
@@ -122,6 +128,7 @@ def __init__(
         super(CosmosDBManagementClient, self).__init__(self.config.credentials, self.config)
 
         client_models = {k: v for k, v in models.__dict__.items() if isinstance(v, type)}
+        self.api_version = '2021-03-01-preview'
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/_cosmos_db_management_client.py:131]: Can you try using this command to generate the sdk :
autorest --version=V2 --use=@microsoft.azure/autorest.python@~4.0.71 --python --python-mode=update --multiapi --keep-version-file --no-async --python-sdks-folder=E:\CosmosDBClientSDKs\azure-sdk-for-python\sdk2 ./readme.md

File: src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/version.py [hunk 5]
```diff
@@ -9,5 +9,5 @@
 # regenerated.
 # --------------------------------------------------------------------------
 
-VERSION = "2.0.0rc3"
+VERSION = "3.0.0"
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/version.py:12]: You can name the sdk version as 3.0.0.rc1 (if this version would have been released it will be called this)

Comment (MehaKaushik) [reply src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/version.py:12]: rc: release candidate (since this is a  preview apiversions)

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 6]
```diff
@@ -139,6 +139,55 @@ def load_arguments(self, _):
...
         c.argument('scope', validator=validate_scope, options_list=['--scope', '-s'], help="Data plane resource path at which this Role Assignment is being granted.")
         c.argument('principal_id', options_list=['--principal-id', '-p'], help="AAD Object ID of the principal to which this Role Assignment is being granted.")
+
+    # Managed Cassandra Cluster
+    with self.argument_context('cassandra-mi cluster create') as c:
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:146]: what is "mi": managed instance ?

Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:146]: yes it is.

Comment (MehaKaushik) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:146]: I think cassandra and cluster can be a single term.

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 7]
```diff
@@ -139,6 +139,55 @@ def load_arguments(self, _):
...
+        c.argument('location', options_list=['--location', '-l'], help="Location", required=True)
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+        c.argument('delegated_management_subnet_id', options_list=['--delegated-management-subnet-id', '-s'], help="Delegated SubnetId", required=True)
+
+    # Managed Cassandra Cluster
+    with self.argument_context('cassandra-mi cluster update') as c:
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+
+    # Managed Cassandra Cluster
+    with self.argument_context('cassandra-mi cluster node status') as c:
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:156]: commands should look like name-of-some-object verb

in should be something like:  cassandra-cluster-node-status show
Is there a way to make cassandra-cluster-node-status shorter ?

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py (hunk 6 above)
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:146]: cassandra-mi is not very clear. Why not cassandra-managed-instance ?

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 8]
```diff
@@ -139,6 +139,55 @@ def load_arguments(self, _):
...
+    with self.argument_context('cassandra-mi cluster delete') as c:
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+
+    # Managed Cassandra Datacenter
+    with self.argument_context('cassandra-mi datacenter create') as c:        
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+        c.argument('data_center_name', options_list=['--data-center-name', '-d'], help="Datacenter Name", required=True)
+        c.argument('data_center_location', options_list=['--data-center-location', '-l'], help="Datacenter Location", required=True)
+        c.argument('delegated_subnet_id', options_list=['--delegated-subnet-id', '-s'], help="Delegated SubnetId", required=True)
+        c.argument('node_count', options_list=['--node-count', '-n'], help="Node Count", required=True)
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:173]: Can you please elaborate the help messages everywhere.

File: src/cosmosdb-preview/azext_cosmosdb_preview/commands.py [hunk 9]
```diff
@@ -100,3 +111,19 @@ def load_command_table(self, _):
         g.command('list', 'list_sql_role_assignments')
         g.show_command('show', 'get_sql_role_assignment')
         g.command('delete', 'delete_sql_role_assignment', confirmation=True)
+
+    with self.command_group('cassandra-mi cluster', cosmosdb_managed_cassandra_cluster_sdk, client_factory=cf_cassandra_cluster, is_preview=True) as g:
+        g.custom_command('create', 'cli_cosmosdb_managed_cassandra_cluster_create')
+        g.custom_command('update', 'cli_cosmosdb_managed_cassandra_cluster_update')
+        g.command('node status', 'fetch_node_status')
+        g.command('list-by-subscription', 'list_by_subscription')
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/commands.py:119]: list will be a single command, in there you can provide different parameter sets for customer. i.e if subscription is provided then you should call the list by subscription sdk method, if rg then the corresponding rg method. If both are provided then throw error.

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py (hunk 6 above)
Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:146]: I think this is a call made by the team and I was given this command by Theo

Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:146]: making the change to cassandra-managed-instance

File: src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/_cosmos_db_management_client.py (hunk 4 above)
Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/_cosmos_db_management_client.py:131]: I have picked the sdk from this location. sdk generated  by PR
https://github.com/Azure/azure-rest-api-specs/pull/12861/checks?check_run_id=1964851992

File: src/cosmosdb-preview/azext_cosmosdb_preview/commands.py (hunk 2 above)
Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/commands.py]: fixed.

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py (hunk 7 above)
Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py]: It has to be cluster node-status.

File: src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/operations/_cassandra_cluster_operations.py [hunk 10]
```diff
@@ -0,0 +1,828 @@
...
+    """
+
+    models = models
+
+    def __init__(self, client, config, serializer, deserializer):
+
+        self._client = client
+        self._serialize = serializer
+        self._deserialize = deserializer
+        self.api_version = "2020-06-01-preview"
```
Comment (AverageDesigner) [inline src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/operations/_cassandra_cluster_operations.py:40]: This has to be fixed.

File: src/cosmosdb-preview/azext_cosmosdb_preview/tests/latest/test_cosmosdb-cassandrami_scenario.py [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+import unittest
+
+from azure_devtools.scenario_tests import AllowLargeResponse
+from azure.cli.testsdk import (ScenarioTest, ResourceGroupPreparer)
+
+
+TEST_DIR = os.path.abspath(os.path.join(os.path.abspath(__file__), '..'))
+
+
+class Cosmosdb_previewScenarioTest(ScenarioTest):
```
Comment (AverageDesigner) [inline src/cosmosdb-preview/azext_cosmosdb_preview/tests/latest/test_cosmosdb-cassandrami_scenario.py:16]: Still working on tests.

**Review by kavskalyan — APPROVED** [review]

**Comment by MehaKaushik** [comment]
Your CI is failing.
Please refer here for linter details. https://azurecliprod.blob.core.windows.net/videos/03%20-%20AzdevBasics.mp4
source: https://github.com/Azure/azure-cli-dev-tools

**Comment by AverageDesigner** [comment]
> Your CI is failing.
> Please refer here for linter details. https://azurecliprod.blob.core.windows.net/videos/03%20-%20AzdevBasics.mp4
> source: https://github.com/Azure/azure-cli-dev-tools

linter changes are in local and will

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 12]
```diff
@@ -142,3 +142,53 @@ def load_arguments(self, _):
...
+    with self.argument_context('cassandra-managed-instance cluster node-status') as c:
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+
+    # Managed Cassandra Cluster
+    with self.argument_context('cassandra-managed-instance cluster show') as c:
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+
+    # Managed Cassandra Cluster
+    with self.argument_context('cassandra-managed-instance cluster delete') as c:
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
```
Comment (evelyn-ys) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:166]: Actually you can create an `cluster_name_type` and apply it to the whole `cluster` command group to avoid duplications.
```suggestion
    from knack.arguments import CLIArgumentType
    cluster_name_type = CLIArgumentType(options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
    # Managed Cassandra Cluster
    with self.argument_context('cassandra-managed-instance cluster') as c:
        c.argument('cluster_name', cluster_name_type)
```

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 13]
```diff
@@ -142,3 +142,53 @@ def load_arguments(self, _):
...
+
+    # Managed Cassandra Datacenter
+    with self.argument_context('cassandra-managed-instance datacenter show') as c:
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+        c.argument('data_center_name', options_list=['--data-center-name', '-d'], help="Datacenter Name", required=True)
+
+    # Managed Cassandra Datacenter
+    with self.argument_context('cassandra-managed-instance datacenter delete') as c:
+        c.argument('cluster_name', options_list=['--cluster-name', '-c'], help="Cluster Name", required=True)
+        c.argument('data_center_name', options_list=['--data-center-name', '-d'], help="Datacenter Name", required=True)
```
Comment (evelyn-ys) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:194]: Same suggestion with `data_center_name`. You can define `data_center_name_type` and apply it to the whole `cassandra-managed-instance datacenter` command group.

File: src/cosmosdb-preview/azext_cosmosdb_preview/commands.py [hunk 14]
```diff
@@ -100,3 +110,18 @@ def load_command_table(self, _):
...
+
+    with self.command_group('cassandra-managed-instance cluster', cosmosdb_managed_cassandra_cluster_sdk, client_factory=cf_cassandra_cluster, is_preview=True) as g:
+        g.custom_command('create', 'cli_cosmosdb_managed_cassandra_cluster_create')
+        g.custom_command('update', 'cli_cosmosdb_managed_cassandra_cluster_update')
+        g.command('node-status', 'fetch_node_status')
+        g.custom_command('list', 'cli_cosmosdb_managed_cassandra_cluster_list')
+        g.show_command('show', 'get')
+        g.command('delete', 'delete', confirmation=True)
+
+    with self.command_group('cassandra-managed-instance datacenter', cosmosdb_managed_cassandra_datacenter_sdk, client_factory=cf_cassandra_data_center, is_preview=True) as g:
```
Comment (evelyn-ys) [inline src/cosmosdb-preview/azext_cosmosdb_preview/commands.py:122]: Do you mind naming it `data-center` instead of `datacenter`?

**Review by evelyn-ys — COMMENTED** [review]
Don't forget to update `HISTORY.rst` and upgrade the version inside `setup.py`.

File: src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/_cosmos_db_management_client.py [hunk 15]
```diff
@@ -106,6 +108,10 @@ class CosmosDBManagementClient(SDKClient):
     :vartype private_link_resources: azure.mgmt.cosmosdb.operations.PrivateLinkResourcesOperations
     :ivar private_endpoint_connections: PrivateEndpointConnections operations
     :vartype private_endpoint_connections: azure.mgmt.cosmosdb.operations.PrivateEndpointConnectionsOperations
+    :ivar cassandra_cluster: CassandraCluster operations
+    :vartype cassandra_cluster: azure.mgmt.cosmosdb.operations.CassandraClusterOperations
+    :ivar cassandra_data_center: CassandraDataCenter operations
+    :vartype cassandra_data_center: azure.mgmt.cosmosdb.operations.CassandraDataCenterOperations
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/vendored_sdks/azure_mgmt_cosmosdb/_cosmos_db_management_client.py:114]: nit: cassandra_cluster should be plural. ( to be fixed in swagger)

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py (hunk 7 above)
Comment (MehaKaushik) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py]: @evelyn-ys any recommendations for the command name here. To me it looks like we are not using a verb, which is not correct. 

---
In reply to: [582376886](https://github.com/Azure/azure-cli-extensions/pull/3035#discussion_r582376886) [](ancestors = 582376886)

File: src/cosmosdb-preview/azext_cosmosdb_preview/_help.py [hunk 16]
```diff
@@ -287,3 +287,38 @@
...
+short-summary: Create a Datacenter in an Azure Managed Cassandra Cluster.
+examples:
+  - name: Create a Managed Cassandra Datacenter in a Cassandra Cluster.
+    text: |
+      az cassandra-managed-instance datacenter create --cluster-name MyCluster --data-center-name MyDataCenter \\
+        --resource-group MyResourceGroup \\
+        --data-center-location westus2 \\
+        --delegated-subnet-id /subscriptions/94d9b402-77b4-4049-b4c1-947bc6b7729b/resourceGroups/My-vnet/providers/Microsoft.Network/virtualNetworks/test-vnet/subnets/test-subnet \\
+        --node-count 3
+"""
```
Comment (MehaKaushik) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_help.py:324]: This doesnt look complete. Please add all the commands and please add elaborate help messages. Please examples for usage as well.

**Review by MehaKaushik — COMMENTED** [review]
Mostly looks good. Tests and help isnt complete though.

File: src/cosmosdb-preview/azext_cosmosdb_preview/_params.py [hunk 17]
```diff
@@ -142,3 +142,69 @@ def load_arguments(self, _):
...
+
+    # Managed Cassandra Cluster
+    for scope in [
+        'cassandra-managed-instance cluster create', 
+        'cassandra-managed-instance cluster update']:        
+        with self.argument_context(scope) as c:
+            c.argument('tags', arg_type=tags_type)
+            c.argument('external_gossip_certificates', options_list=['--external-gossip-certificates', '-e'], help="A list of certificates that the managed cassandra data center's should accept.")
+            c.argument('initial_cassandra_admin_password', options_list=['--initial-cassandra-admin-password', '-i'], help="The intial password to be configured when a cluster is created for authentication_method Cassandra. If none is specified, the password will be cassandra.")
+            c.argument('delegated_management_subnet_id', options_list=['--delegated-management-subnet-id', '-s'], help="The resource if of a subnet where the ip address of the cassandra management server will be allocated. This subnet must have connectivity to the delegated_subnet_id subnet of each data center.")
```
Comment (evelyn-ys) [inline src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:164]:
```suggestion
            c.argument('delegated_management_subnet_id', options_list=['--delegated-management-subnet-id', '-s'], help="The resource id of a subnet where the ip address of the cassandra management server will be allocated. This subnet must have connectivity to the delegated_subnet_id subnet of each data center.")
```

Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_params.py:164]: fixed.

File: src/cosmosdb-preview/azext_cosmosdb_preview/_help.py (hunk 16 above)
Comment (AverageDesigner) [reply src/cosmosdb-preview/azext_cosmosdb_preview/_help.py:324]: yes adding as many as possible. Once this PR is merged and the team uses it for bug bash we plan to add more examples and help messages.
